### PR TITLE
Reconcile labels from discovered EC2 instances

### PR DESCRIPTION
**Pull request by atburke**
This PR adds the `labelReconciler` to the discovery service, which periodically reconciles the labels it receives from discovered EC2 instances with the labels of the corresponding SSH servers stored in the auth server.

[RFD](https://github.com/gravitational/teleport/blob/master/rfd/0105-ec2-tags.md)

### CODE REVIEW

File: lib/srv/discovery/reconciler.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	return batchSize
+}
+
+func (r *labelReconciler) run(ctx context.Context) {
+	ticker := r.cfg.clock.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.Chan():
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/srv/discovery/reconciler.go]: Per [this comment](https://github.com/gravitational/teleport/pull/25281#discussion_r1186051228), I ended up not adding `UpsertServerInfos` due to not really having backend support for batch operations.

[the remark this answers is not in the document]
Comment (atburke) [reply lib/srv/discovery/reconciler.go]: I changed the ticker to be just a `clock.After`. That should guarantee at least a one second gap between upserts.

File: lib/srv/server/ec2_watcher.go [hunk 2]
```diff
@@ -81,6 +89,34 @@ func ToEC2Instances(insts []*ec2.Instance) []EC2Instance {
 
 }
 
+// ServerInfos creates a ServerInfo resource for each discovered instance.
+func (i *EC2Instances) ServerInfos() ([]types.ServerInfo, error) {
+	serverInfos := make([]types.ServerInfo, 0, len(i.Instances))
+	for _, instance := range i.Instances {
+		name := i.AccountID + "-" + instance.InstanceID
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/srv/server/ec2_watcher.go:97]: This is running on the discovery service, not on discovered nodes. AccountID and InstanceID here will come from the EC2 API, so access to the metadata server won't be an issue.

File: lib/srv/server/ec2_watcher.go [hunk 3]
```diff
@@ -81,6 +90,34 @@ func ToEC2Instances(insts []*ec2.Instance) []EC2Instance {
 
 }
 
+// ServerInfos creates a ServerInfo resource for each discovered instance.
+func (i *EC2Instances) ServerInfos() ([]types.ServerInfo, error) {
```
Comment (gabrielcorado) [inline lib/srv/server/ec2_watcher.go:94]: Suggestion: Moving the "covert" logic to `services/server_info.go`. This makes it easier to keep consistency across different providers (if/when added) and better for testing (which might be necessary to cover here too). We follow this when creating databases from the AWS SDK structs ([here is an example](https://github.com/gravitational/teleport/blob/master/lib/services/database.go#L693) and [its tests](https://github.com/gravitational/teleport/blob/master/lib/services/database_test.go#L710-L731)).

**Review by gabrielcorado — APPROVED** [review]
LGTM.

File: lib/srv/server/ec2_watcher.go [hunk 4]
```diff
@@ -62,12 +63,20 @@ type EC2Instances struct {
 // discovered.
 type EC2Instance struct {
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/srv/server/ec2_watcher.go:64]: I assume so, but you'd have to ask @lxea.

File: lib/srv/server/ec2_watcher.go (hunk 3 above)
Comment (atburke) [reply lib/srv/server/ec2_watcher.go:94]: Definitely on board, but doing this right now leads to an import cycle between `lib/services` and `lib/labels`.

**Comment by atburke** [comment]
Friendly ping @espadolini

**Comment by atburke** [comment]
Friendly ping @espadolini @rosstimothy

**Comment by r0mant** [comment]
@espadolini Can you please review this when you get a chance?

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**